### PR TITLE
Add `gamepad_stick_input` example

### PR DESCRIPTION
**Pull request by mfdorst**
# Objective

In pursuit of #3691 I felt that it would be handy to have a way to visualize the gamepad stick input.

## Solution

Add an example that shows gamepad stick input.

gamepad_stick_input3

The dark gray regions in the center and the outer border are the default deadzones.

### CODE REVIEW

**Comment by mfdorst** [comment]
Still adding some stuff. I want to mark the deadzone regions with boxes.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]

File: examples/input/gamepad_stick_input.rs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            .unwrap();
+        xform.translation.x = x * WINDOW_SIZE / 2.0;
+        xform.translation.y = y * WINDOW_SIZE / 2.0;
+        text.sections[0].value = format!("({:6.3}, {:6.3})", x, y);
+    }
+}
+
+fn setup(mut commands: Commands, asset_server: Res<AssetServer>) {
+    // Spawn cameras
+    commands.spawn_bundle(OrthographicCameraBundle::new_2d());
```
[the remark this answers is not in the document]
Comment (mfdorst) [reply examples/input/gamepad_stick_input.rs]: Lol mistakes were made.

File: examples/input/gamepad_stick_input.rs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        })
+        .insert_resource(ClearColor(DEADZONE_COLOR))
+        .add_plugins(DefaultPlugins)
+        .add_startup_system(setup)
+        .add_system(update_position)
+        .run();
+}
+
+fn update_position(
+    mut xform_query: Query<&mut Transform, With<Crosshair>>,
```
Comment (rparrett) [inline examples/input/gamepad_stick_input.rs:32]: For consistency with other examples, I think `crosshair_query` would be better here.

File: examples/input/gamepad_stick_input.rs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    mut text_query: Query<&mut Text, With<CoordinateText>>,
+    gamepads: Res<Gamepads>,
+    axes: Res<Axis<GamepadAxis>>,
+) {
+    let mut xform = xform_query.single_mut();
```
Comment (rparrett) [inline examples/input/gamepad_stick_input.rs:37]: For consistency with other examples, I think `let mut transform` would be better here.

File: examples/input/gamepad_stick_input.rs [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    // Spawn camera
+    commands.spawn_bundle(OrthographicCameraBundle::new_2d());
+
+    // Spawn crosshair
+    let texture = asset_server.load("textures/crosshair.png");
+    commands
+        .spawn_bundle(SpriteBundle {
+            texture,
+            sprite: Sprite {
+                custom_size: Some(Vec2::new(CROSSHAIR_SIZE, CROSSHAIR_SIZE)),
```
Comment (rparrett) [inline examples/input/gamepad_stick_input.rs:67]: When specifying the size of something that's explicitly square, `Vec2::splat(CROSSHAIR_SIZE)` is nice.

File: examples/input/gamepad_stick_input.rs [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    let deadzone_box_midpoint = deadzone_midpoint * WINDOW_SIZE / 2.0;
+    let livezone_box_size = livezone_size * WINDOW_SIZE / 2.0;
+    let deadzone_box_size = deadzone_size * WINDOW_SIZE / 2.0;
+    // For text placement
+    let livezone_lower_left_corner = (livezone_box_midpoint - livezone_box_size) / 2.0;
+
+    // Spawn livezone box
+    commands.spawn_bundle(SpriteBundle {
+        sprite: Sprite {
+            custom_size: Some(Vec2::new(livezone_box_size, livezone_box_size)),
```
Comment (rparrett) [inline examples/input/gamepad_stick_input.rs:95]: Same as above

File: examples/input/gamepad_stick_input.rs [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+            color: LIVEZONE_COLOR,
+            ..Default::default()
+        },
+        transform: Transform::from_xyz(livezone_box_midpoint, livezone_box_midpoint, 0.0),
+        ..Default::default()
+    });
+    // Spawn deadzone box
+    commands.spawn_bundle(SpriteBundle {
+        sprite: Sprite {
+            custom_size: Some(Vec2::new(deadzone_box_size, deadzone_box_size)),
```
Comment (rparrett) [inline examples/input/gamepad_stick_input.rs:105]: Same as above

**Review by rparrett — COMMENTED** [review]
This seems like a good addition to me. We don't expose gamepad settings anywhere else and it looks nice!

**Comment by alice-i-cecile** [comment]
Can we add a circle with radius 1 to verify that the joystick never leaves it?

File: examples/input/gamepad_stick_input.rs (hunk 6 above)
Comment (mfdorst) [reply examples/input/gamepad_stick_input.rs:105]: Oh cool I wasn't aware of that. Thanks

**Comment by mfdorst** [comment]
Not sure how to render a circle but I'll try and figure it out.

**Comment by mockersf** [comment]
you can't currently without adding a lot of unrelated complexity (or using a picture of a circle)

**Comment by alice-i-cecile** [comment]
I think either "picture of a circle" or just leaving it would be fine for this initial example.

**Comment by mfdorst** [comment]
The circle thing adds a bit of complexity because the way the outside deadzone is currently shown is that the clear color is the deadzone color and the livezone is shown with a lighter gray sprite. If I want to show the circle I'd probably to have the livezone be a light gray circle and then render four sprites around the edges for the outside deadzone. I can do that but I think it might be better not to in the interest of simplicity. Let me know what you think.

**Comment by mfdorst** [comment]
Also, all of this will need to change when #3464 lands, because it will no longer be possible to visualize the deadzone areas  unless we add a way to get the raw unscaled input (incidentally I think that would be handy because then I could show two crosshairs, one for scaled and one for unscaled, but admittedly I can't see a use-case for an actual game).

**Comment by ickk** [comment]
> Can we add a circle with radius 1 to verify that the joystick never leaves it?

I don't think this is something that controllers generally guarantee. In hardware joysticks are usually a pair of linear variable resistors for each of the x & y axes, with the radius only being constrained physically by the cutout in the plastic form, not electrically. You can verify this by testing your own controllers with something like [gamepad-tester.com](https://gamepad-tester.com/). 

Here's an xbox controllers which has an interesting rounded-square shape because the variable resistors actually max out before being restricted when positioned at the extremes directly along the x/y axes, though it does become constrained by the circular cutout elsewhere: 
image

Many controllers have differently shaped cutouts, some circular and some octagonal, and there may be some that happen to not limit the radius of the stick at all beyond the extremes of each axis (so x: 1.0, y: 1.0 may be valid/attainable).

**Comment by mfdorst** [comment]
@rparrett oh I'm sorry, I wasn't aware of the spam issue. I will fix that.

**Review by alice-i-cecile — APPROVED** [review]

**Comment by Nilirad** [comment]
This PR adds a new example. Adding module and item level doc comments, as described in:

- #3951 and
- #4438,

would be really useful to those who will browse examples.

**Comment by alice-i-cecile** [comment]
@rparrett could I get your approval on this? This is a super cool example that I'd like to include.

**Comment by rparrett** [comment]
This PR needs to be updated due to #4519 being merged.

```diff
index 8f9e2cfed..bf8bb4f01 100644
--- a/examples/input/gamepad_stick_input.rs
+++ b/examples/input/gamepad_stick_input.rs
@@ -39,10 +39,10 @@ fn update_position(
     for gamepad in gamepads.iter() {
         // We only use input from the left stick.
         let x = axes
-            .get(GamepadAxis(*gamepad, GamepadAxisType::LeftStickX))
+            .get(GamepadAxis::new(*gamepad, GamepadAxisType::LeftStickX))
             .unwrap();
         let y = axes
-            .get(GamepadAxis(*gamepad, GamepadAxisType::LeftStickY))
+            .get(GamepadAxis::new(*gamepad, GamepadAxisType::LeftStickY))
             .unwrap();
         transform.translation.x = x * WINDOW_SIZE / 2.0;
         transform.translation.y = y * WINDOW_SIZE / 2.0;
```

**Comment by alice-i-cecile** [comment]
Oh and we also merged #3730, so we can add a unit circle indicator :D

**Comment by rparrett** [comment]
I do think it would be a good idea to rework this so that it's window-size-agnostic / just displays the thing in the center of the screen. (Thinking about https://bevyengine.org/examples/)

> If I want to show the circle I'd probably to have the livezone be a light gray circle and then render four sprites around the edges for the outside deadzone.

Drawing the edges that way shouldn't be necessary. In order of increasing Z you can spawn

- Big dead zone square
- Live zone square
- Circle
- Little dead zone square

**Comment by alice-i-cecile** [comment]
Closing in favor of the rebased https://github.com/bevyengine/bevy/pull/6052 <3 Excellent work with the prototyping here though; thanks!

**Pull request closed without merging**